pcoredump.1
.TH PCOREDUMP "1" "Sep 2007" "pcoredump 0.1.2" "User Commands"
.SH NAME
pcoredump \- take a core dump of a running program without terminating it

.SH SYNOPSIS
.B pcoredump
.RB [ \-f | \-\-force\-limit=
.IR LIMIT ]
.RB [ \-i | \-\-io\-prio=
.IR IOPRIO ]
.RB [ \-s | \-\-sched\-prio=
.IR PRIO ]
.RB [ \-o | \-\-oom\-adjustment=
.IR ADJ ]
PID [\fIPID...\fR]

.SH DESCRIPTION
.PP
Take a "live" core dump from the process specified by PID.  The whole
process, including all of it's threads, will be stopped and cloned.
Then, it's VM will be copied to the clone. Finally, the original
process will be started again from the point where it was stopped, and
the clone process will dump core and exit.  The resulting core dump
file will have the same structure as the core file generated by the
normal mechanism (for example, due to handling SIGSEGV), so gdb(1) may
be used to analyze it.
.PP
Running with no OPTION and PID, pcoredump outputs help text and
exits.  The following options are supported:

.TP
\fB\-f LIMIT\fR, \fB\-\-force-limit=LIMIT\fR
Force the file size limit for the core dump being produced. The syntax
of LIMIT is similar to ulimit(1) builtin of bash(1), i.e. LIMIT must
be in bytes or "unlimited", which means no limit. The number of bytes must
be greater than 0. The CAP_SYS_RESOURCE capability is required to
force the limit; otherwise, the value of LIMIT will be silently
ignored.

.TP
\fB\-i PRIO\fR, \fB\-\-io-prio=PRIO\fR
Set the I/O priority for the clone process. PRIO must be from 0 to 7,
just as for the ionice(1) command.  Changing I/O scheduling class is not
supported.

.TP
\fB\-o ADJ\fR, \fB\-\-oom-adjustment=ADJ\fR
Set the out-of-memory killer adjustment of the clone process. The
value of ADJ must be from OOM_DISABLE to OOM_ADJUST_MAX (see
include/linux/oom.h to see numerical values). The value of ADJ
specifies the likeliness of this process being killed if the system
runs out of memory.

.TP
\fB\-s PRIO\fR, \fB\-\-sched-prio=PRIO\fR
Set scheduling niceness of clone process. Value of PRIO must be from
-20 to 19, and the meaning of this value is the same as for nice(1)
command.

.TP
\fB PID\fR
The process of which to take the core dump.  If more than one PID is
specified, a core dump of each process is taken in sequence.

.PP
By default, the clone process will be created with an ultra-low
priority values and core file size limit will be inherited from the
process being dumped. Core file will be created as specified by the
pattern in /proc/sys/kernel/core_pattern.  The current directory of
the dumped process is used, which is often "/" for daemons.

.PP
It's impossible to take a live core dump of a kernel thread, take a
live coredump of a process which is exiting (performing an exit(3)
call) or handling a fatal signal, or from the process which is already
doing another "live" core dump.  Only one live coredump may be running
at a time in the system; other live coredumps will be blocked until
the running one is completed.

.SH AUTHOR
The pcoredump utility and initial version of this manual page is written
by Dmitry Antipov <antipov@ru.mvista.com>, and Corey Minyard
<cminyard@mvista.com> has make this page even better (and more readable).

.SH "REPORTING BUGS"
Report bugs to <source@mvista.com>.

.SH "SEE ALSO"
.BR nice(1),
.BR ionice(1),
.BR gdb(1),
.BR ulimit(1)

.SH COPYRIGHT
Copyright \(co 2007 Montavista Software, Inc.


pcoredump.c
#ifndef _GNU_SOURCE
#define _GNU_SOURCE 1
#endif

#include "config.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#include <errno.h>
#include <getopt.h>
#include <error.h>
#include <sys/resource.h>
#include <sys/ptrace.h>

#ifndef PTRACE_LIVEDUMP
#define PTRACE_LIVEDUMP 0x4221
#endif
#ifndef PT_LIVEDUMP
#define PT_LIVEDUMP PTRACE_LIVEDUMP
#endif

#ifdef HAVE_LINUX_OOM_H
#include <linux/oom.h>
#else
#define OOM_DISABLE -17
#define OOM_ADJUST_MIN -16
#define OOM_ADJUST_MAX 15
#endif

#ifdef HAVE_LINUX_LIVEDUMP_H
#include <linux/livedump.h>
#else
struct livedump_param {
  int sched_nice;
  int io_prio;
  int oom_adj;
  unsigned long core_limit;
};
#endif

long
parse_numeric (char *arg)
{
  char *p;
  long val;

  val = strtol (arg, &p, 10);
  if (val == LONG_MAX)
    error (1, 0, "invalid number '%s'", arg);
  else if (*p || errno)
    error (1, errno, "not a number '%s'", arg);
  return val;
}

void
usage (void)
{
  fprintf (stderr, "usage: %s [options] pid [pid..]\n", program_invocation_short_name);
  fprintf (stderr, "Options are:\n");
  fprintf (stderr, "  -f, --force-limit=[NR,unlimited]\n"
	           "        force maximum core file size to NR bytes or unlimited\n");
  fprintf (stderr, "  -i, --io-priority=prio       set I/O priority\n");
  fprintf (stderr, "  -o, --oom-adjustment=adj     set OOM killer adjustment\n");
  fprintf (stderr, "  -s, --sched-nice=nice        set scheduling niceness\n");
  fprintf (stderr, "Report bugs to <%s>.\n", PACKAGE_BUGREPORT);
  exit (1);
}

int
main (int argc, char *argv[])
{
  int opt, changed;
  long val;
  struct livedump_param param = {0, 0, 0, 0};
  struct option pcoredump_options[] =
    {
      {"force-limit", required_argument, 0, 'f'},
      {"io-priority", required_argument, 0, 'i'},
      {"oom-adjustment", required_argument, 0, 'o'},
      {"sched-nice", required_argument, 0, 's'},
      {0, 0, 0, 0}
    };

  if (argc < 2)
    usage ();
  changed = 0;

  while ((opt = getopt_long (argc, argv, "f:i:o:s:",
			     pcoredump_options, NULL)) != -1)
    {
      if (opt == 'f')
	{
	  if (!strcmp (optarg, "unlimited"))
	    param.core_limit = RLIM_INFINITY;
	  else
	    {
	      val = parse_numeric (optarg);
	      if (val < 1 || val > RLIM_INFINITY)
		error (1, 0, "invalid core file size limit %ld - must be [1:%ld]\n", 
		       val, RLIM_INFINITY);
	      param.core_limit = val;
	    }
	  changed++;
	}
      else if (opt == 'i')
	{
	  val = parse_numeric (optarg);
	  if (val < 0 || val > 7)
	    error (1, 0, "invalid IO priority %ld - "
		   "must be in range [0:7]", val);
	  param.io_prio = val;
	  if (val)
	    changed++;
	}
      else if (opt == 'o')
	{
	  val = parse_numeric (optarg);
	  if (val < OOM_DISABLE || val > OOM_ADJUST_MAX)
	    error (1, 0, "invalid OOM adjustment %ld - must be in range "
		   "[%d:%d]", val, OOM_DISABLE, OOM_ADJUST_MAX);
	  param.oom_adj = val;
	  if (val)
	    changed++;
	}
      else if (opt == 's')
	{
	  val = parse_numeric (optarg);
	  if (val < -20 || val > 19)
	    error (1, 0, "invalid scheduling niceness %ld - "
		   "must be in range [-20:19]", val);
	  param.sched_nice = val;
	  if (val)
	    changed++;
	}
      else
	usage ();
    }

  if (optind < argc)
    {
      while (optind < argc)
	{
	  long pid;

	  pid = parse_numeric (argv[optind]);
	  if (pid < 0)
	    error (1, 0, "invalid pid %lu\n", pid);
	  if (ptrace (PT_LIVEDUMP, pid, NULL, changed ? &param : NULL) < 0)
	    error (1, errno, "failed to dump core of %ld", pid);
	  optind++;
	}
    }
  else
    usage ();
  return 0;
}
